web/ui/react-app/src/pages/graph/PanelList.tsx

```diff
@@ -1,219 +1,181 @@
       (acc, metric) => {
         return metric === query ? acc : [...acc, metric]; // Prevent adding query twice.
       },
       [query]
     );
-    localStorage.setItem('history', JSON.stringify(extendedItems.slice(0, 50)));
-    this.updatePastQueries();
-  };
-
-  updateURL() {
-    const query = encodePanelOptionsToQueryString(this.state.panels);
-    window.history.pushState({}, '', query);
-  }
-
-  handleOptionsChanged = (id: string, options: PanelOptions) => {
-    const updatedPanels = this.state.panels.map(p => (id === p.id ? { ...p, options } : p));
-    this.setState({ panels: updatedPanels }, this.updateURL);
+    setLocalStorageHistoryItems(extendedItems.slice(0, 50));
   };
 
-  addPanel = () => {
-    const { panels } = this.state;
-    const nextPanels = [
+  const addPanel = () => {
+    callAll(setPanels, updateURL)([
       ...panels,
       {
         id: generateID(),
         key: `${panels.length}`,
         options: PanelDefaultOptions,
       },
-    ];
-    this.setState({ panels: nextPanels }, this.updateURL);
+    ]);
   };
 
-  removePanel = (id: string) => {
-    this.setState(
-      {
-        panels: this.state.panels.reduce<PanelMeta[]>((acc, panel) => {
-          return panel.id !== id ? [...acc, { ...panel, key: `${acc.length}` }] : acc;
-        }, []),
-      },
-      this.updateURL
-    );
-  };
+  return (
+    <>
+      {panels.map(({ id, options }) => (
+        <Panel
+          onExecuteQuery={handleExecuteQuery}
+          key={id}
+          options={options}
+          onOptionsChanged={opts =>
+            callAll(setPanels, updateURL)(panels.map(p => (id === p.id ? { ...p, options: opts } : p)))
+          }
+          removePanel={() =>
+            callAll(setPanels, updateURL)(
+              panels.reduce<PanelMeta[]>(
+                (acc, panel) => (panel.id !== id ? [...acc, { ...panel, key: `${acc.length}` }] : acc),
+                []
+              )
+            )
+          }
+          useLocalTime={useLocalTime}
+          metricNames={metrics}
+          pastQueries={queryHistoryEnabled ? historyItems : []}
+          pathPrefix={pathPrefix}
+        />
+      ))}
+      <Button color="primary" onClick={addPanel}>
+        Add Panel
+      </Button>
+    </>
+  );
+};
+
+const PanelsWithStatus = withStatusIndicator<PanelListProps>(PanelListContent);
+
+const PanelList: FC<RouteComponentProps & PathPrefixProps> = ({ pathPrefix }) => {
+  const [metricNames, setMetricNames] = useState<string[]>([]);
+  const [timeDriftError, setTimeDriftError] = useState();
+  const [fetchMetricsError, setFetchMetricsError] = useState();
+  const [useLocalTime, setUseLocalTime] = useLocalStorage('use-local-time', false);
+  const [enableQueryHistory, setEnableQueryHistory] = useLocalStorage('enable-query-history', false);
+
+  useEffect(() => {
+    (async () => {
+      const metricsResponse = await fetch(`${pathPrefix}/api/v1/label/__name__/values`, { cache: 'no-store' });
```

Comment:
Meanwhile in master the code changed to also include `credentials: 'same-origin'` in the fetch options (also in useFetch).